tools/astarte_export/lib/fetchdata/queries/queries.ex

```diff
@@ -0,0 +1,249 @@
+          tag: "database_error"
+        )
+
+        {:error, :database_error}
+
+      {:error, %Xandra.ConnectionError{} = err} ->
+        Logger.error("database connection error:#{inspect(err)}.",
+          tag: "database_connection_error"
+        )
+
+        {:error, :database_connection_error}
+    end
+  end
+
+  def retrieve_individual_datastreams(
+        conn,
+        realm,
+        device_id,
+        interface_id,
+        endpoint_id,
+        path,
+        data_type,
+        options
+      ) do
+    individual_datastream_statement = """
+    SELECT #{data_type}, reception_timestamp FROM  #{realm}.individual_datastreams WHERE device_id=? AND
+      interface_id=? AND endpoint_id=? AND path=?
+    """
+
+    params = [{"uuid", device_id}, {"uuid", interface_id}, {"uuid", endpoint_id}, {"text", path}]
+
+    page_options = get_page_options(options)
+
+    options = [uuid_format: :binary, timestamp_format: :integer] ++ page_options
+
+    with {:ok, result} <- Xandra.execute(conn, individual_datastream_statement, params, options) do
+      {:ok, result}
+    else
+      {:error, %Xandra.Error{message: message}} ->
+        Logger.error("database error: #{inspect(message)}.",
+          realm: realm,
+          device_id: device_id,
+          tag: "database_error"
+        )
+
+        {:error, :database_error}
+
+      {:error, %Xandra.ConnectionError{} = err} ->
+        Logger.error("database connection error:#{inspect(err)}.",
+          realm: realm,
+          device_id: device_id,
+          tag: "database_connection_error"
+        )
+
+        {:error, :database_connection_error}
+    end
+  end
+
+  def retrieve_object_datastream_value(conn, realm, storage, device_id, path, options) do
+    object_datastream_statement = """
+      SELECT * from #{realm}.#{storage} where device_id=? AND path=?
+    """
+
+    params = [{"uuid", device_id}, {"text", path}]
+
+    page_options = get_page_options(options)
+
+    options = [uuid_format: :binary, timestamp_format: :integer] ++ page_options
+
+    with {:ok, result} <-
+           Xandra.execute(conn, object_datastream_statement, params, options) do
+      {:ok, result}
+    else
+      {:error, %Xandra.Error{message: message}} ->
+        Logger.error("database error: #{inspect(message)}.",
+          realm: realm,
+          device_id: device_id,
+          tag: "database_error"
+        )
+
+        {:error, :database_error}
+
+      {:error, %Xandra.ConnectionError{} = err} ->
+        Logger.error("database connection error: #{inspect(err)}.",
+          realm: realm,
+          device_id: device_id,
+          tag: "database_connection_error"
+        )
+
+        {:error, :database_connection_error}
+    end
+  end
+
+  defp get_page_options(options) do
+    case Keyword.get(options, :page_size) do
+      nil ->
+        []
+
+      page_size ->
+        case Keyword.get(options, :paging_state) do
+          nil ->
+            [page_size: page_size]
```

Comment:
Honesty I can't understand the whole `page_size` thing. Also wouldn't be easier for you to just use https://hexdocs.pm/xandra/Xandra.html#stream_pages!/4 ?

Reply:
It looks, There is not much diffrence b/w execute and stream_pages

Moreover for our usecase, it is better to have execute itself.

Reply:
I realized, in Queries.ex there is no need of guarding or validation of page_size options. As, we are doing those validation in fetch and export module.

Made this more generic now